scene/3d/mesh_instance_3d.cpp

```diff
@@ -117,6 +117,12 @@ void MeshInstance3D::set_mesh(const Ref<Mesh> &p_mesh) {
 		mesh->connect(CoreStringNames::get_singleton()->changed, callable_mp(this, &MeshInstance3D::_mesh_changed));
 		_mesh_changed();
 		set_base(mesh->get_rid());
+		int surface_count = mesh->get_surface_count();
+		for (int surface_index(0); surface_index < surface_count; ++surface_index) {
+			if (surface_override_materials[surface_index].is_valid()) {
+				RS::get_singleton()->instance_set_surface_override_material(get_instance(), surface_index, surface_override_materials[surface_index]->get_rid());
+			}
+		}
```

Comment:
This should move to be inside of ``_mesh_changed()`` so it is called everytime the mesh changes. For example, if a new surface is added to the mesh, you want this to be called, so it needs to be inside ``_mesh_changed()``